<!doctype html>
<html lang="en">

<head>
  <meta http-equiv="x-ua-compatible" content="ie=edge">
  <title> The Art of Machine Learning </title>
  <meta name="description" content="machine learning, data science, data visualization, deep learning">
  <meta name="viewport" content="width=device-width, initial-scale=1">

  <meta property="og:title" content="The Art of Machine Learning">
  <meta property="og:type" content="website">
  <meta property="og:url" content="http://yinniyu.github.io/">

  <meta property="og:image" content="http://chalk.nielsenramon.com/assets/og-image.jpg">


  <link rel="apple-touch-icon" href="/assets/apple-touch-icon.png">
  <link href="http://chalk.nielsenramon.com/feed.xml" type="application/rss+xml" rel="alternate" title="Chalk Last 10 blog posts" />

  

  
    <link rel="stylesheet" href="/assets/light.css">

  
</head>

<body>
  <main>
    <div class="grid grid-centered">
      <div class="grid-cell">
        <nav class="header-nav appear">
  <a href="/" class="header-logo" title="BlogTitle">The Art of Machine Learning</a>
  <ul class="header-links">
    
    
      <li>
        <a href="about">About Me
        </a>
      </li>
    
    
    
      <li>
        <a href="https://github.com/yinniyu" rel="noreferrer noopener" target="_blank" title="GitHub">
          <span class="icon icon-github"></span>
        </a>
      </li>
    
    
    
  
      <li>
        <a href="https://linkedin.com/in/jenny-yu-b495243" rel="noreferrer noopener" target="_blank" title="LinkedIn">
          <span class="icon icon-linkedin"></span>
        </a>
      </li>
    
  </ul>
</nav>

        <ul class="article-list">
        <li class="article-list-item appear">
          <a href="/posts/tumor_rna" title="Sparse PCA and XGboost">
            <h5>Sparse PCA and XGboost for predicting tumor type</h5>
              <figure><img src="posts/images/rna_seq/xgb_multi_tree.png" alt="xgb" ></figure>
              <p>The Pan-Cancer intiative at <a href="https://portal.gdc.cancer.gov/">The Cancer Genome Atlas (TCGA) Research Network </a>  
              has analyzed numerous types of tumor using different technologies to profile the genomic, transcriptomic and epigenomic landscapes.
              This body of work represents an approach to study cancers not only based on the organ of origin, but also the broader biomarker 
              profiles [1]. For this post, high-throughput gene expression data are used to build a classifier for 5 tumor types: breast cancer, lung cancer,
                colorectal cancer, kidney cancer and prostate cancer. The public data was provided and maintained by TCGA, the raw RNA sequence data were 
              normalized to Fragment per kilobase transcript per million mapped reads (FPKM) to adjust for differences of transcript lengths.
              The normalized data are compiled and hosted by <a href="https://archive.ics.uci.edu/ml/datasets/gene+expression+cancer+RNA-Seq#">UCI Machine Learning Repository</a>. 
            </p>
              <span class="icon icon-arrow-right"></span>
            
          </a>
      
      <div class="article-list-footer">
        <span class="article-list-date">
          December 4th, 2017
        </span>
        <span class="article-list-divider">-</span>
        <span class="article-list-minutes">
          15 minute read
          
        </span>
        <span class="article-list-divider">-</span>
        <div class="article-list-tags">
        <span class="article-list-divider">R, Sparse PCA, XGBoost, Cancer Research </span>
      </div>
    </li>
            
          
        <li class="article-list-item appear">
          <a href="/posts/WMH" title="White matter hyperintensities segmentation">
            <h5>Segmentation of white matter hyperintensities using 3D U-net </h5>
              <figure><img src="posts/images/WMH/mask overlay.png" ></figure>
              <p>White Matter Hyperintensities (aka., leukoaraiosis) is often linked to high risk of stroke and dementia in 
            older patients[1]. While image segmentation is critical for timely diagnosis and evaluation of treatments, 
            automated segmentation of medical images remains a challenging task. In this post, a deep learning network called U-net 
            is evaluated on brain MRIs for segmentation of White Matter Hyperintensities (WMH). The dataset was obtained from the 
            WMH Segmentation Challenge, which's organized by UMC Utrecht, VU Amsterdam, and NUHS Singapore. The goal is to train the 
            deep learning model to generate binary mask that corresponds to WMH region of the brain MRI. This kind of deep learning 
            model has great potential for automated medical diagnostic applications. 
            In fact, commercial models have been developed for heart and breast imaging [2].</p>
             <span class="icon icon-arrow-right"></span>
        
      </a>
      
      <div class="article-list-footer">
        <span class="article-list-date">
          November 5, 2017
        </span>
        <span class="article-list-divider">-</span>
        <span class="article-list-minutes">
          15 minute read
          
        </span>
        <span class="article-list-divider">-</span>
        <div class="article-list-tags">
        <span class="article-list-divider">Python, Convolutional Neural Net, Segmentation </span>
      </div>
    </li>
          
          
          
     <li class="article-list-item appear">
      <a href="/posts/quora" title="Detecting duplicate question pair">
        <h5>
           Designing a classifier to detect duplicate questions
          <p>For this Kaggle competition that took place 6 months ago, the goal is to compile a model to identify if a pair of questioins is asking the same thing. Quora provided 400K+ question pairs for the training set, and the final test data set has 2,345,796 question pairs (that's alot of data!). Many Kagglers have tried techniques such as Xgboost and feature extraction such as TF-IDF, ratio of matching words, and weighted word2vec. I wanted to learn how recurrent neural network works, and explore its potential in solving this problem and others (i.e., genomic functional prediction). 
            The final ranking for this submission was top 37% on Kaggle's private LB.</p>
          <span class="icon icon-arrow-right"></span>
        </h5>
      </a>
      
      <div class="article-list-footer">
        <span class="article-list-date">
          November 2, 2017
        </span>
        <span class="article-list-divider">-</span>
        <span class="article-list-minutes">
          
          
            10 minute read
          
        </span>
        <span class="article-list-divider">-</span>
        <div class="article-list-tags">
        <span class="article-list-divider">Python, Keras, LSTM, WordVec </span>
      </div>
    </li>
  
  
    <li class="article-list-item appear">
      <a href="/posts/taxi" title="Predicting trip duration for NY taxis">
        <h5>
          Data visualization and predicting taxi ride time
          <p>This post will cover some data visualization techniques for working with high dimensional dataset. 
            Then I will talk about using Random Forest and Caret to predict trip duration with cluster computing.</p>
          <span class="icon icon-arrow-right"></span>
        </h5>
      </a>
      
      <div class="article-list-footer">
        <span class="article-list-date">
          October 2, 2017
        </span>
        <span class="article-list-divider">-</span>
        <span class="article-list-minutes">
          
          
            10 minute read
          
        </span>
        <span class="article-list-divider">-</span>
        <div class="article-list-tags">
        <span class="article-list-divider">R, caret, Regression, ggmap, ggplot </span>
      </div>
    </li>
  
    
  


        <footer class="footer appear">

</footer>

      </div>
    </div>
  </main>
  
  <script>
    window.ga=function(){ga.q.push(arguments)};ga.q=[];ga.l=+new Date;
    ga('create','UA-28631876-6','auto');ga('send','pageview')
  </script>
  <script src="https://www.google-analytics.com/analytics.js" async></script>


<script src="/assets/vendor.js"></script>



  <script src="/assets/webfonts.js"></script>




  <script src="/assets/scrollappear.js"></script>



<script src="/assets/application.js"></script>


</body>
</html>
